src/App.tsx

```diff
@@ -17,18 +53,36 @@ export const App: React.FC = () => {
             <h1 className="title">Todos:</h1>
 
             <div className="block">
-              <TodoFilter />
+              <TodoFilter
+                filter={filter}
+                query={query}
+                onFilterChange={setFilter}
+                onQueryChange={setQuery}
+              />
             </div>
 
             <div className="block">
-              <Loader />
-              <TodoList />
+              {!todos.length
+                ? <Loader />
+                : (
+                  <TodoList
+                    todos={visibleTodos}
+                    selectedTodo={selectedTodo}
+                    onSelectTodo={(id: number) => setSelectedTodoId(id)}
+                  />
```

Comment:
Don't use inliners
You have two options:

1. pass `setState` callback, this function never changes, so you don't have useless re-renders.
2. create a specific function that takes value and triggers `setState` function with that value. But you need to wrap this function with useCallback (and you will get render optimization) - that is what you actually did and it's okay

If you pass `(value) => setState(value)`- it will cause re-renders of the TodoList